dss-plugin-kml-format/ : Add read capability for KMZ, add unit tests in Java for both KML and KMZ

tdesfont opened the pull request:
Started from the existing KML plugin, working on adding read of KMZ files and added tests using Gradle.

tdesfont committed:
Init Gradle for KML/KMZ format plugin

tdesfont committed:
Add gradle tests and build

tdesfont committed:
Add tests in Gradle. One test working for the KML read, the DSS object mock is to be validated using hamcrest in the second

tdesfont committed:
Add the tests on KML and KMZ. Passed unit tests on KMZ parsing.

tdesfont committed:
Remove unused comments.

tdesfont commented:
**NB**: Still have to create a test project, check documentation and code, will add the links here soon

tdesfont committed:
Fix guava version

tdesfont committed:
build.gradle: Change gradle file

tdesfont committed:
Add Junit test for KMZ

tdesfont committed:
KMLFormat.java: Include the KMZ processing in the KML format
KMZFormatTest.java: Removed test that is not passing (Should understand why it does not pass)

tdesfont commented:
**Status on that PR:** The unit tests on the KMZ are failing, I don't know exactly why since the extracted content looks ok. I have put the KML and KMZ in the same file so that the KML format in DSS can handle both formats now. Thanks JF for the guava fix and the help with gradle. I will update you as soon as the PR can be reviewed.

tdesfont committed:
KMZFormatTest.java: Add several tests for KMZ format, added a TODO on the failing test case

tdesfont committed:
KMZFormatTest.java: Fix dumb mistake in the test case

tdesfont commented:
I set up a test project on dku17: http://dku17.dataiku.com:30800/projects/PLUGINTESTKMLKMZ/flow/

jfyuen commented on build.gradle:
```diff
@@ -0,0 +1,110 @@
+sourceSets {
+    main {
+        java {
+            srcDirs = ['java-formats/kml-format_kml/src']
```
We already talked about it, it seems weird to me that the src files are in `java-formats/kml-format_kml/src` and tests files in `tests/java/unit`.
Have a look [here](https://docs.gradle.org/current/userguide/organizing_gradle_projects.html) at how gradle expects directories

MayeulRousselet replied:
Sorry to disturb maven purists (I am one of them), but that's how things works with Java components in plugins.

jfyuen commented on build.gradle:
```diff
@@ -0,0 +1,110 @@
+ */
+
+plugins {
+	id 'java'	
```
May want to use the [java-library](https://docs.gradle.org/current/userguide/java_library_plugin.html) plugin instead

jfyuen commented on build.gradle:
```diff
@@ -0,0 +1,110 @@
+task copyDependencies(type: Copy) {
+    from configurations.runtimeOnlyDeps
+    from configurations.remoteCompileDeps
+    into 'java-lib'
```
Also extends the `clean` task to delete that directory:
```
clean.doLast {
    delete 'java-lib'
    delete 'lib'
}
```

jfyuen reviewed (changes requested):
If moving to gradle, you can also remove the `build.xml` for Ant.

Only looked at the gradle build, not the implementation. 
You may also want to replace all those `System.out.Println()` by a logger.

tdesfont commented:
Hey @jfyuen @MayeulRousselet you can start the review, I am making sure that test flows are available today, do not hesitate to ping me

tdesfont commented:
@jfyuen I just saw your comments above, thanks !

tdesfont committed:
Minor update in the README.md

tdesfont committed:
KMLFormat.java: Refactor common input stream processing
Tests: Add missing filename of InputStreamWithContextInfo
Unit Tests are running

tdesfont committed:
Clean state

tdesfont committed:
Refactoring of the tests

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,12 @@
+# KML & KMZ format extractor
+
+- Expose a KML Parser
```
And KMZ?

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,12 @@
+
+## Testing the plugin
+
+To build and test the plugin using Gradle, do:
```
```suggestion
To test the plugin using Gradle, do:
./gradlew test
```

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,12 @@
+- Expose a KML Parser
+- Run tests of the project.
+
+## Testing the plugin
```
The paragraph is not very clear, the title is "Testing the plugin" whereas what you describe is the build & release process. Make a "build the plugin" paragraph, and after a "Run the unit tests" paragraph.

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,12 @@
+./gradlew build
+./gradlew dist
+
+Once done, a zip of the plugin is created in the dist folder. This archive will be the one used by DSS.
```
```suggestion
Once done, a zip of the plugin is created in the dist folder. This archive will be the one used by DSS to install the plugin.
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/formats/kml/KMLFormat.java:
```diff
@@ -246,14 +147,14 @@ public void setWarningsContext(WarningsContext warningsContext) {
-
-    private static Logger logger = Logger.getLogger("dku");
+    protected static DKULogger logger = DKULogger.getLogger("dku");
+    // private static Logger logger = Logger.getLogger("dku");
```
remove comment

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/formats/kml/KMLFormat.java:
```diff
@@ -246,14 +147,14 @@ public void setWarningsContext(WarningsContext warningsContext) {
     }
-
-    private static Logger logger = Logger.getLogger("dku");
+    protected static DKULogger logger = DKULogger.getLogger("dku");
```
Change the visibility from protected to private.

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+
+    public KMLParser() {}
+
+    public static Logger logger = Logger.getLogger("dku");
```
Change the visibility from public to private and put the declaration at the end of the class.

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+    public static Logger logger = Logger.getLogger("dku");
+
+    public Element getFirstNodeByTagName(Element parent, String name){
+        logger.info("Children of" + parent);
```
remove the logging

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        else return (Element)nl.item(0);
+    }
+
+    public String getTextContent(Node e){return e.getTextContent();}
```
Remove this method and inline the calls (can be done in IntelliJ by selecting the method, right click "Refactor" > inline method)

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        Node childNode = getFirstNodeByTagName(e, childNodeName);
+        if (childNode != null) {
+            String txt = getTextContent(childNode);
+            if (txt != null) r.put(cf.column(columnName), txt);
```
```suggestion
            if (txt != null) {
                 r.put(cf.column(columnName), txt);
            }
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+
+    private void parsePlacemark(Node node, ProcessorOutput out, ColumnFactory cf, RowFactory rf) throws Exception {
+        Element e = (Element)node;
+        //assert(node.getLocalName().equals("Placemark"));
```
remove comment

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        }
+    }
+
+    private void parsePlacemark(Node node, ProcessorOutput out, ColumnFactory cf, RowFactory rf) throws Exception {
```
The formatting of this method is ugly (it is a remain of the old plugin), remove the blank spaces.

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        cf.column("id");
+        putAttrValueIfExists(cf, r, "id", e, "id");
+
+        cf.column("geom");
```
Redondant, cf.column("geom") will be called in the block below. To remove

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        Row r = rf.row();
+
+        putContentIfExistsInChild(cf, r, "name", e, "name");
+        cf.column("id");
```
Redondant, cf.column("id") will be called in putAttrValueIfExists. To remove

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+            if (pointNode != null) {
+                Element coordsElt = getFirstNodeByTagName(pointNode, "coordinates");
+                // Mandatory
+                String coordsTxt = getTextContent(coordsElt);
```
As getFirstNodeByTagName can return a null value, test if coordsElt != null

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+
+        cf.column("geom");
+
+        {
```
Anonymous blocks in Java are to be avoided.
For a better readability, create a private method "extractAndSetPoint"

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+                String coordsTxt = getTextContent(coordsElt);
+                String[] chunks = coordsTxt.split(",");
+
+                Coords coords = new Coords(Double.parseDouble(chunks[1]), Double.parseDouble(chunks[0]));
```
Same here, 
```suggestion
                if(chunks != null && chunks.size() >=2) {
                      Coords coords = new Coords(Double.parseDouble(chunks[1]), Double.parseDouble(chunks[0]));
                      r.put(cf.column("geom"), coords.toWKT());
                 }
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+            }
+        }
+
+        {
```
Extract this block into a extractAndSetLine

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+            if (linestringNode != null) {
+                Element coordsElt = getFirstNodeByTagName(linestringNode, "coordinates");
+                // Mandatory
+                String coordsTxt = getTextContent(coordsElt);
```
Same, test for if coordsElt is not null

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+                // Mandatory
+                String coordsTxt = getTextContent(coordsElt);
+                logger.info("Parse linestring: " + coordsTxt);
+                String[] points = StringUtils.splitByWholeSeparator(coordsTxt,  " ");
```
Same, test if points is not null

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+
+                List<String> pointsStr = new ArrayList<>();
+                for (String point : points) {
+                    if (StringUtils.isBlank(point)) continue;
```
```suggestion
                    if (StringUtils.isBlank(point)){
                        continue;
                   }
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+                    if (StringUtils.isBlank(point)) continue;
+                    logger.info("POINT: --" + point + "--");
+                    String[] chunks = point.split(",");
+                    pointsStr.add(chunks[1] + " " + chunks[0]);
```
```suggestion
if(chunks != null && chunks.size() >=2) {
                    pointsStr.add(chunks[1] + " " + chunks[0]);
               }
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+                    String[] chunks = point.split(",");
+                    pointsStr.add(chunks[1] + " " + chunks[0]);
+                }
+                r.put(cf.column("geom"), "LINESTRING(" + StringUtils.join(pointsStr, ",") + ")");
```
Here, test that pointsStr has more than 2 elements. As my grandpa said, "Can't create a line out of one point man!"

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        }
+
+        {
+            Element extendedDataElt = getFirstNodeByTagName(e, "ExtendedData");
```
Saaaame, create a method call extractAndSetExtendedData

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,140 @@
+        {
+            Element extendedDataElt = getFirstNodeByTagName(e, "ExtendedData");
+            if (extendedDataElt != null) {
+                NodeList nl = extendedDataElt.getElementsByTagName("Data");
```
Test for nullability of nl

MayeulRousselet commented on tests/java/unit/com/dataiku/dss/formats/kml/KMLFormatTest.java:
```diff
@@ -0,0 +1,198 @@
+    };
+
+    private CustomFormatInput kmlFormatInput;
+    private ColumnFactory colFactory = new StreamColumnFactory();
```
As ColumnFactory and RowFactory maintains an internal state (the columns created and the rows), best is to init both member in setupDSS(), as we do for kmlFormatInput

MayeulRousselet commented on tests/java/unit/com/dataiku/dss/formats/kml/KMLFormatTest.java:
```diff
@@ -0,0 +1,198 @@
+
+
+    @Test
+    public void testReadFile() throws IOException {
```
This test has no assert. Either add one or remove the test

tdesfont committed:
build.gradle: Use java library plugin
README.md: Better readability
KMLFormat.java: Changed logger to private
KMLParser.java: Declare logger at the end of file, removed unused logging, created methods extractAndSet{} for Point, LineString and ExtendedData, added checks on null values
KMLFormatTest.java: Moved instantiation of colFactory and rowFactory in setUpDSS, removed testReadFile()
sample_kml.kml: Removed this unused file

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,21 @@
+- Expose a KML/KMZ parser
+- Run tests
+
+## Building the plugin
```
Build the plugin

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,21 @@
+
+It will create a zip of the plugin in the dist folder. DSS uses this zip to install the plugin.
+
+## Testing the plugin
```
Test the plugin

MayeulRousselet commented on README.md:
```diff
@@ -0,0 +1,21 @@
+# KML & KMZ format extractor
+
+- Expose a KML/KMZ parser
```
Could be more precise: What the plugin does?

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+    public KMLParser() {}
+
+    public Element getFirstNodeByTagName(Element parent, String name){
+        for (int i=0; i < parent.getChildNodes().getLength(); i++){
```
Remove this logging too

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+                logger.info("Parse linestring: " + coordsTxt);
+                String[] points = StringUtils.splitByWholeSeparator(coordsTxt,  " ");
+                List<String> pointsStr = new ArrayList<>();
+                for (String point : points) {
```
test for points nullability

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+            // Mandatory
+            if (coordsElt != null){
+                String coordsTxt = coordsElt.getTextContent();
+                String[] chunks = coordsTxt.split(",");
```
Check if coordsTxt is not null

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+            if (coordsElt != null){
+                String coordsTxt = coordsElt.getTextContent();
+                logger.info("Parse linestring: " + coordsTxt);
+                String[] points = StringUtils.splitByWholeSeparator(coordsTxt,  " ");
```
Check if coordsTxt is not null

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    private void parsePlacemark(Node node, ProcessorOutput out, ColumnFactory cf, RowFactory rf) throws Exception {
+        Element e = (Element)node;
```
Test if node can be casted to Element
```suggestion
        if( node instanceof Element) {
              Element e = (Element)node;
              ...
       }
```

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+            NodeList nl = extendedDataElt.getElementsByTagName("Data");
+            if (nl != null){
+                for (int i = 0; i < nl.getLength(); i++) {
+                    Element dataElt = (Element)nl.item(i);
```
Idem, test if nb.item(i) can be casted to Element

MayeulRousselet commented on java-formats/kml-format_kml/src/com/dataiku/dss/utils/KMLParser.java:
```diff
@@ -0,0 +1,145 @@
+            Node childNode = containerNode.getChildNodes().item(i);
+            logger.info("Check child " + childNode);
+            if (childNode instanceof Element) {
+                Element childElt = (Element)childNode;
```
Idem, test if childNode can be casted to Element

tdesfont committed:
Review build process of the plugin and zip

tdesfont committed:
Update README.md

tdesfont committed:
Corrections from review

MayeulRousselet reviewed (changes requested):
Faut updater le numero de version dans le MAkefile et le plugin.json
